driver-core/src/test/unit/com/mongodb/internal/connection/TestConnectionPoolListener.java

```diff
@@ -28,52 +28,115 @@
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
 
 class TestConnectionPoolListener implements ConnectionPoolListener {
 
     private final List<Object> events = new ArrayList<Object>();
+    private final Lock lock = new ReentrantLock();
+    private final Condition condition = lock.newCondition();
+    private volatile Class<?> waitingForEventClass;
+    private volatile int waitingForEventCount;
 
     public List<Object> getEvents() {
-        return events;
+        lock.lock();
+        try {
+            return new ArrayList<Object>(events);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public <T> void waitForEvent(final Class<T> eventClass, final int count, final long time, final TimeUnit unit)
+            throws InterruptedException, TimeoutException {
+        lock.lock();
+        try {
+            if (waitingForEventClass != null) {
+                throw new IllegalStateException("Already waiting for events of class " + waitingForEventClass);
+            }
+            waitingForEventClass = eventClass;
+            waitingForEventCount = count;
+            if (containsEvent(eventClass, count)) {
```

Comment:
Early exit if the condition is already satisfied.  We don't wait to wait for it if it's already been signaled.